### PR TITLE
fix: handle falsy node I/O properly

### PR DESCRIPTION
**Pull request by MichelBartels**
As discovered in #67, outputting 0 is equivalent to outputting nothing which shouldn't be the case. The validation also misses when an input is set to 0.
This PR fixes this at two points in the code which fix the issue for me. However, it might also occur somewhere else which I have missed.

I have run the tests manually and they are all passing.

### CODE REVIEW

**Comment by silvanocerza** [comment]
@MichelBartels can you add a test that reproduces the issue maybe?

**Comment by MichelBartels** [comment]
Ok, I have added two tests. I am not sure if there's a better place to put `test_falsy_connection` than `test_pipeline.py`.

**Comment by ZanSara** [comment]
Seems like a good place to me :+1:

File: test/pipelines/unit/test_validation_pipeline_io.py [hunk 1]
```diff
@@ -177,6 +177,14 @@ def test_validate_pipeline_input_only_expected_input_is_present():
...
         pipe.run({"comp1": {"value": 1}, "comp2": {"value": 2}})
 
+def test_validate_pipeline_input_only_expected_input_is_present_falsy():
+    pipe = Pipeline()
+    pipe.add_component("comp1", Double())
+    pipe.add_component("comp2", Double())
+    pipe.connect("comp1", "comp2")
+    with pytest.raises(ValueError, match="The input value of comp2 is already sent by node comp1"):
+        pipe.run({"comp1": {"value": 1}, "comp2": {"value": 0}})
+
```
Comment (ZanSara) [inline test/pipelines/unit/test_validation_pipeline_io.py:188]: Let's add a test to make sure that if a component outputs a zero to the end of the pipeline, it's actually sent to the pipeline output and it doesn't get discarded/filtered anywhere

File: test/pipelines/unit/test_pipeline.py [hunk 2]
```diff
@@ -317,3 +317,20 @@ def test_from_dict_without_connection_receiver():
...
+    @component
+    class A:
+        @component.output_types(y=int)
+        def run(self, x: int):
+            return {"y": 0}
+    @component
+    class B:
+        @component.output_types(y=int)
+        def run(self, x: int):
+            return {"y": x}
```
Comment (silvanocerza) [inline test/pipelines/unit/test_pipeline.py:331]: You can use the `component_class` factory to do this.
https://github.com/deepset-ai/canals/blob/main/canals/testing/factory.py#L9

File: test/pipelines/unit/test_validation_pipeline_io.py [hunk 3]
```diff
@@ -177,6 +177,14 @@ def test_validate_pipeline_input_only_expected_input_is_present():
...
```
[the remark this answers is not in the document]
Comment (MichelBartels) [reply test/pipelines/unit/test_validation_pipeline_io.py:187]: Ok, done. This actually uncovered another bug.

File: test/pipelines/unit/test_pipeline.py (hunk 2 above)
Comment (MichelBartels) [reply test/pipelines/unit/test_pipeline.py]: Thanks, I have changed that.

File: test/pipelines/unit/test_validation_pipeline_io.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZanSara) [reply test/pipelines/unit/test_validation_pipeline_io.py:187]: Nice! Good to know they weren't as useless as they seemed :smile:

File: canals/pipeline/validation.py [hunk 4]
```diff
@@ -66,10 +66,11 @@ def _validate_input_sockets_are_connected(graph: networkx.MultiDiGraph, input_va
     for node, sockets in valid_inputs.items():
         for socket in sockets:
             inputs_for_node = input_values.get(node, {})
+            print(inputs_for_node)
```
Comment (ZanSara) [inline canals/pipeline/validation.py:69]:
```suggestion
```

File: test/pipelines/unit/test_validation_pipeline_io.py [hunk 5]
```diff
@@ -177,6 +177,20 @@ def test_validate_pipeline_input_only_expected_input_is_present():
...
+def test_validate_pipeline_falsy_input():
+    pipe = Pipeline()
+    pipe.add_component("comp", Double())
+    pipe.run({"comp": {"value": 0}})
+    with pytest.raises(ValueError, match="Missing input: comp.value"):
+        pipe.run({"comp": {}})
```
Comment (ZanSara) [inline test/pipelines/unit/test_validation_pipeline_io.py:193]: Sorry I'm being a bit pedantic here :sweat_smile:  Let's split these two tests

```suggestion
def test_validate_pipeline_falsy_input_present():
    pipe = Pipeline()
    pipe.add_component("comp", Double())
    assert pipe.run({"comp": {"value": 0}}) == {"comp": {"value": 0}}
        
def test_validate_pipeline_falsy_input_missing():
    pipe = Pipeline()
    pipe.add_component("comp", Double())
    with pytest.raises(ValueError, match="Missing input: comp.value"):
        pipe.run({"comp": {}})
```

**Review by ZanSara — COMMENTED** [review]
Very last details and LGTM

**Review by silvanocerza — APPROVED** [review]
